lib/dry/schema/macros/key.rb

```diff
@@ -52,19 +62,27 @@ def to_ast
 
         private
 
-        def extract_type_spec(*args)
+        def extract_type_spec(*args, nullable: false)
           type_spec = args[0]
 
           if type_spec.kind_of?(Schema::Processor) || type_spec.is_a?(Symbol) && type_spec.to_s.end_with?(QUESTION_MARK)
             type_spec = nil
           end
 
-          predicates = type_spec ? args[1, -1] : args
+          predicates = Array(type_spec ? args[1, -1] : args)
 
-          type(type_spec) if type_spec
+          if type_spec
+            type(nullable && !type_spec.is_a?(::Array) ? [:nil, type_spec] : type_spec)
+            type_predicate = infer_type_predicate
+            predicates << type_predicate unless predicates.include?(type_predicate)
```

Comment:
Would we want to put the type_predicate onto the front of the list of predicates, rather than onto the end? If we were writing this by hand, after all, we'd probably be putting the type predicate at the beginning?